MovingBlocks/Terasology
# feat(discord-rpc): improve the system

### Contains
This improvement aims to fix bugs in the system.
And, it improves communication between the game and Discord RPC to provide full access to its features.

![image](https://user-images.githubusercontent.com/20463031/100094928-d16e2600-2e6a-11eb-97a6-5ea37f121182.png)
_[Source: **Discord Developer Documentation**](https://discord.com/developers/docs/rich-presence/how-to#updating-presence-update-presence-payload-fields)_

Currently, the system supports `state` that can be changed dynamically using `DiscordRPCSubSystem#setState`.
And, `details` is also supported but its value is fixed as `Name: PlayerName`. It can't be changed.

This PR is going to implement other features such as `partySize`, `partyMax`, etc...

#### Discussion Points
- Turning the `default` to be dynamic to have different details from different gameplay such as `Currently in JoshariasSurvival`, etc...

- Since the engine has a networking infrastructure I'm thinking that we can implement the secrets (`join` and `spectate`)
in a way to provide the modules the ability to receive events like `DiscordAskToJoinEvent` and `DiscordSpectateEvent`.
Instead of each module dealing with it in a raw way.

_**Comment about what do you think of the points above?**_ 🤔 

#### Bugs
- **Game detection is broken**
The game assumes the player is in the lobby while in reality, the player is in-game.
- **Not thread-safe**
The Discord RPC connection is being handled in a separate thread to avoid conflicts with the game.
The system isn't thread-safe which might result in unexpected behavior.

### How to test
- Fetch the PR commits
`git fetch origin pull/4271/head:<localBranchName>`
- Checkout to the PR's branch
`git checkout <localBranchName>`
- Run the game and make sure you have `Discord Rich Presence` turned on.
- Join a game to trigger your game state from `In Lobby` to `In Game`

**NOTE:**
In terms of unit testing, I'm not sure how I would implement it for this system. 😅 

### Outstanding before merging
- [x] Fix: Game detection is broken (solved by #4274)
- [x] Fix: Not thread-safe (47bae74)
- [x] Feat: Ability to change `details` (48d1502)
- [ ] Feat: Implement party size
- [ ] Feat: Implement secrets
- [ ] Test the RPC
- - [ ] Single Player
- - [ ] Multiplayer
